File: src/main/java/Reader.java

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

class Reader
{
	public static Data read(final String fileToParse)
	{
		Data data = null;
		BufferedReader fileReader = null;
		final String delimiter = ";";
		try
		{
			String line = null;
			String[] tokens = null;
			Charge charge = null;
			VacationType[] vacationTypeList = null;
			
			fileReader = new BufferedReader(new FileReader(fileToParse));
			
			/* Step 1. Read the number of period */
			line = fileReader.readLine(); // Read line #1
			tokens = line.split(delimiter);
			charge = new Charge(Integer.parseInt(tokens[1]));
			
			/* Step 2. Read the charge for each period */
			line = fileReader.readLine(); // Read line #2
			line = fileReader.readLine(); // Read line #3
			tokens = line.split(delimiter);
			for(int period = 0; period < charge.getNumberOfPeriods(); ++period)
			{
				charge.set(period, Integer.parseInt(tokens[period]));
			}
			
			/* Step 3. Read the number of vacation types */
			line = fileReader.readLine(); // Read line #4
			line = fileReader.readLine(); // Read line #5
			tokens = line.split(delimiter);
			vacationTypeList = new VacationType[Integer.parseInt(tokens[1])];
			for(int typeIndex = 0; typeIndex < vacationTypeList.length; ++typeIndex)
			{
				vacationTypeList[typeIndex] = new VacationType();
			}
			
			/* Step 4. Read data for each vacation type */
			line = fileReader.readLine(); // Read line #6
			for(VacationType vacationType : vacationTypeList)
			{
				line = fileReader.readLine(); // Read next line
				tokens = line.split(delimiter);
				vacationType.rangeMin = Integer.parseInt(tokens[1]);
				vacationType.rangeMax = Integer.parseInt(tokens[2]);
				vacationType.startMin = Integer.parseInt(tokens[3]);
				vacationType.stopMax = Integer.parseInt(tokens[4]);
				vacationType.breakDurationMin = Integer.parseInt(tokens[5]);
				vacationType.breakDurationMax = Integer.parseInt(tokens[6]);
				vacationType.startToBreakMin = Integer.parseInt(tokens[7]);
				vacationType.breakToStopMin = Integer.parseInt(tokens[8]);
				vacationType.limit = Integer.parseInt(tokens[9]);
				vacationType.cost = Integer.parseInt(tokens[10]);
			}

			data = new Data(charge, vacationTypeList);
		}
		catch(Exception e)
		{
			e.printStackTrace();
		}
		finally
		{
			try
			{
				fileReader.close();
			}
			catch (IOException e)
			{
				e.printStackTrace();
			}
		}
		return data;
	}
}

File: cplusplus/reader.cpp
#include "reader.h"
#include "csv_file.h"

#include <vector>
#include <fstream>
#include <sstream>

using namespace std;

Data * Input::read(const char * filename)
{
    Data * data = NULL;
    try
    {
        CsvFile file;
        ifstream stream(filename);
        file << stream;
        stream.close();
        unsigned int numberOfPeriods, numberOfVacationType, integers[10];
        vector< unsigned int > load;
        vector< VacationType > vacationTypes;

        /* Step 1. Read the number of periods */
        istringstream(file.get(0, 1)) >> numberOfPeriods;

        /* Step 2. Read the charge for each period */
        for(unsigned int period = 0; period < numberOfPeriods; ++period)
        {
            istringstream(file.get(2, period)) >> integers[0];
            load.push_back(integers[0]);
        }

        /* Step 3. Read the number of vacation types */
        istringstream(file.get(4, 1)) >> numberOfVacationType;

        /* Step 4. Read data for each vacation type */
        for(unsigned int typeIndex = 0; typeIndex < numberOfVacationType; ++typeIndex)
        {
            istringstream(file.get(6 + typeIndex,  1)) >> integers[0];
            istringstream(file.get(6 + typeIndex,  2)) >> integers[1];
            istringstream(file.get(6 + typeIndex,  3)) >> integers[2]; --(integers[2]);
            istringstream(file.get(6 + typeIndex,  4)) >> integers[3]; --(integers[3]);
            istringstream(file.get(6 + typeIndex,  5)) >> integers[4];
            istringstream(file.get(6 + typeIndex,  6)) >> integers[5];
            istringstream(file.get(6 + typeIndex,  7)) >> integers[6];
            istringstream(file.get(6 + typeIndex,  8)) >> integers[7];
            istringstream(file.get(6 + typeIndex,  9)) >> integers[8];
            istringstream(file.get(6 + typeIndex, 10)) >> integers[9];
            vacationTypes.push_back(
                VacationType(
                    integers[0],
                    integers[1],
                    integers[2],
                    integers[3],
                    integers[4],
                    integers[5],
                    integers[6],
                    integers[7],
                    integers[8],
                    integers[9]
                )
            );
        }

        data = new Data(load.begin(), load.end(), vacationTypes.begin(), vacationTypes.end());
    }
    catch(...)
    {
        if(data != NULL)
        {
            delete data;
            data = NULL;
        }
    }
    return data;
}
